src/workerd/api/streams/standard.c++

```diff
@@ -1536,37 +1542,59 @@ void ReadableStreamBYOBRequest::respond(jsg::Lock& js, int bytesWritten) {
       JSG_REQUIRE(bytesWritten > 0,
                   TypeError,
                   "The bytesWritten must be more than zero while the stream is open.");
-      impl.readRequest->respond(js, bytesWritten);
+      if (impl.readRequest->respond(js, bytesWritten)) {
+        // The read request was fulfilled, we need to invalidate.
+        shouldInvalidate = true;
+      } else {
+        // The response did not fulfill the minimum requirements of the read.
+        // We do not want to invalidate the read request and we need to update the
+        // view so that on the next read the view will be properly adjusted.
+        impl.updateView(js);
+      }
     }
-    pull = true;
-  }
-
-  KJ_DEFER(invalidate(js));
-  if (pull) {
     impl.controller->pull(js);
+    if (shouldInvalidate) {
+      invalidate(js);
+    }
   }
 }
 
 void ReadableStreamBYOBRequest::respondWithNewView(jsg::Lock& js, jsg::BufferSource view) {
   auto& impl = JSG_REQUIRE_NONNULL(maybeImpl,
                                     TypeError,
                                     "This ReadableStreamBYOBRequest has been invalidated.");
-  bool pull = false;
   if (!impl.controller->canCloseOrEnqueue()) {
     JSG_REQUIRE(view.size() == 0,
                  TypeError,
                  "The view byte length must be zero after the stream is closed.");
+    KJ_ASSERT(impl.readRequest->isInvalidated());
+    invalidate(js);
   } else {
-    JSG_REQUIRE(view.size() > 0,
-                 TypeError,
-                 "The view byte length must be more than zero while the stream is open.");
-    impl.readRequest->respondWithNewView(js, kj::mv(view));
-    pull = true;
-  }
+    bool shouldInvalidate = false;
+    if (impl.readRequest->isInvalidated()) {
```

Comment:
`::respond()` checks the consumer count here -- should this code path as well? If there's a reason not, worth a comment.

Reply:
Yes, good catch